### PR TITLE
Add defined.cr

### PR DESCRIPTION
**Pull request by wyhaines**
## Link
https://github.com/wyhaines/defined.cr

## Checklist
* [x] - Shard is at least 30 days old.
* [x] - Shard has CI implemented.
* [x] - Shard has daily/weekly periodic builds (ideally with Crystal latest and nightly).

I have been using this package heavily in some other libraries. It provides facilities to enable conditional compilation based on whether or not a constant exists, version numbers match requirements, or environment variables are defined (or not). Recent documentation should cover the use cases. There is spec coverage, and specs are setup to run nightly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,69 +12,69 @@ Search [Crystal Shards](https://crystalshards.xyz) or follow announcements [Crys
...
+- [Community](#community)
+  - [Unofficial](#unofficial)
+- [Resources](#resources)
+  - [Official Documentation Translations](#official-documentation-translations)
+- [Services and Apps](#services-and-apps)
+- [Tools](#tools)
+  - [DevOps](#devops)
+  - [Editor Plugins](#editor-plugins)
+  - [LSP Language Server Protocol Implementations](#lsp-language-server-protocol-implementations)
+  - [Shell plugins](#shell-plugins)
```
Comment (Blacksmoke16) [inline README.md:77]: Probably should revert this change.

File: README.md [hunk 2]
```diff
@@ -342,6 +342,7 @@ Contributions are welcome. Please take a quick look at the [contribution guideli
  * [crystal-binary_parser](https://github.com/DanSnow/crystal-binary_parser) - Binary parser
  * [crystal-web-framework-stars](https://github.com/isaced/crystal-web-framework-stars) - Web frameworks for Crystal, most starred on Github
  * [crz](https://github.com/dhruvrajvanshi/crz) - Functional programming library
+ * [defined](https://github.com/wyhaines/defined.cr) - macros for conditional compilation based on constant definitions, version requirements, or environment variable settings.
```
Comment (Blacksmoke16) [inline README.md:345]:
```suggestion
 * [defined](https://github.com/wyhaines/defined.cr) - macros for conditional compilation based on constant definitions, version requirements, or environment variable settings
```

File: README.md (hunk 1 above)
Comment (wyhaines) [reply README.md:77]: Whoa.  What the heck? I made a fresh fork, and added one line and one line only.  *sigh*

I'm going to kill this PR completely, and start over in a new branch, and see if I can figure out what the heck happened there.
